# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [45]:
# Your code goes here
import pandas as pd

url1 = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
url2 = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv"
url3 = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv"

df1 = pd.read_csv(url1)
df2 = pd.read_csv(url2)
df3 = pd.read_csv(url3)

df1.columns = df1.columns.str.lower().str.strip().str.replace(" ", "_")
df2.columns = df2.columns.str.lower().str.strip().str.replace(" ", "_")
df3.columns = df3.columns.str.lower().str.strip().str.replace(" ", "_")

# unimos (concatenar filas)
df = pd.concat([df1, df2, df3], ignore_index=True)

# vemos estructura inicial
df.head()

,customer,st,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount,state
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934,NaN
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935,NaN
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247,NaN
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344,NaN
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323,NaN


In [46]:
df[["state", "st"]].isna().mean()

state    0.414444
st       0.828806
dtype: float64

In [47]:
df[["state", "st"]].nunique()

state    5
st       8
dtype: int64

In [48]:
df["state"] = df["state"].fillna(df["st"])
df = df.drop(columns=["st"])
df.head()


,customer,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount,state
0,RB50392,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934,Washington
1,QZ44356,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935,Arizona
2,AI49188,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247,Nevada
3,WW63253,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344,California
4,GA49547,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323,Washington


In [50]:
#limpiamos customer lifetime value
df["customer_lifetime_value"] = (
    df["customer_lifetime_value"]
    .astype(str)
    .str.replace("%", "")
)

df["customer_lifetime_value"] = pd.to_numeric(df["customer_lifetime_value"], errors="coerce")

In [51]:
#limpiamos income y otros
df["income"] = pd.to_numeric(df["income"], errors="coerce")
df["monthly_premium_auto"] = pd.to_numeric(df["monthly_premium_auto"], errors="coerce")
df["total_claim_amount"] = pd.to_numeric(df["total_claim_amount"], errors="coerce")

In [52]:
#limpiamos texto
for col in df.select_dtypes(include="object").columns:
    df[col] = df[col].str.strip().str.lower()

In [53]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12074 entries, 0 to 12073
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   9137 non-null   object 
 1   gender                     9015 non-null   object 
 2   education                  9137 non-null   object 
 3   customer_lifetime_value    9130 non-null   float64
 4   income                     9137 non-null   float64
 5   monthly_premium_auto       9137 non-null   float64
 6   number_of_open_complaints  2067 non-null   object 
 7   policy_type                9137 non-null   object 
 8   vehicle_class              9137 non-null   object 
 9   total_claim_amount         9137 non-null   float64
 10  state                      9137 non-null   object 
dtypes: float64(4), object(7)
memory usage: 1.0+ MB


,customer,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount,state
0,rb50392,NaN,master,NaN,0.0,1000.0,1/0/00,personal auto,four-door car,2.704934,washington
1,qz44356,f,bachelor,697953.59,0.0,94.0,1/0/00,personal auto,four-door car,1131.464935,arizona
2,ai49188,f,bachelor,1288743.17,48767.0,108.0,1/0/00,personal auto,two-door car,566.472247,nevada
3,ww63253,m,bachelor,764586.18,0.0,106.0,1/0/00,corporate auto,suv,529.881344,california
4,ga49547,m,high school or below,536307.65,36357.0,68.0,1/0/00,personal auto,four-door car,17.269323,washington


In [55]:
df["number_of_open_complaints"].value_counts().head()
df["number_of_open_complaints"].unique()[:10]

array([ 0.,  2.,  1.,  3.,  5.,  4., nan])

In [56]:
# falta limpiar complaints

df["number_of_open_complaints"] = (
    df["number_of_open_complaints"]
    .astype(str)
    .str.split("/").str[1]
)

df["number_of_open_complaints"] = pd.to_numeric(df["number_of_open_complaints"], errors="coerce")

In [57]:
df.duplicated().sum()

np.int64(2939)

In [58]:
df.isna().mean().sort_values(ascending=False)

number_of_open_complaints    1.000000
gender                       0.253354
customer_lifetime_value      0.243830
customer                     0.243250
education                    0.243250
income                       0.243250
monthly_premium_auto         0.243250
policy_type                  0.243250
vehicle_class                0.243250
total_claim_amount           0.243250
state                        0.243250
dtype: float64

In [59]:
df.head()

,customer,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount,state
0,rb50392,NaN,master,NaN,0.0,1000.0,NaN,personal auto,four-door car,2.704934,washington
1,qz44356,f,bachelor,697953.59,0.0,94.0,NaN,personal auto,four-door car,1131.464935,arizona
2,ai49188,f,bachelor,1288743.17,48767.0,108.0,NaN,personal auto,two-door car,566.472247,nevada
3,ww63253,m,bachelor,764586.18,0.0,106.0,NaN,corporate auto,suv,529.881344,california
4,ga49547,m,high school or below,536307.65,36357.0,68.0,NaN,personal auto,four-door car,17.269323,washington


# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [61]:
# Your code goes here
import pandas as pd

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv"
df = pd.read_csv(url)

df.head()

,unnamed:_0,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,...,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type,month
0,0,DK49336,Arizona,4809.216960,No,Basic,College,2011-02-18,Employed,M,...,9,Corporate Auto,Corporate L3,Offer3,Agent,292.800000,Four-Door Car,Medsize,A,2
1,1,KX64629,California,2228.525238,No,Basic,College,2011-01-18,Unemployed,F,...,1,Personal Auto,Personal L3,Offer4,Call Center,744.924331,Four-Door Car,Medsize,A,1
2,2,LZ68649,Washington,14947.917300,No,Basic,Bachelor,2011-02-10,Employed,M,...,2,Personal Auto,Personal L3,Offer3,Call Center,480.000000,SUV,Medsize,A,2
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,2011-01-11,Employed,M,...,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A,1
4,4,QA50777,Oregon,9025.067525,No,Premium,Bachelor,2011-01-17,Medical Leave,F,...,7,Personal Auto,Personal L2,Offer1,Branch,707.925645,Four-Door Car,Medsize,A,1


In [62]:
df = df.drop(columns=["unnamed:_0"], errors="ignore")
df.head()

,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,income,...,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type,month
0,DK49336,Arizona,4809.216960,No,Basic,College,2011-02-18,Employed,M,48029,...,9,Corporate Auto,Corporate L3,Offer3,Agent,292.800000,Four-Door Car,Medsize,A,2
1,KX64629,California,2228.525238,No,Basic,College,2011-01-18,Unemployed,F,0,...,1,Personal Auto,Personal L3,Offer4,Call Center,744.924331,Four-Door Car,Medsize,A,1
2,LZ68649,Washington,14947.917300,No,Basic,Bachelor,2011-02-10,Employed,M,22139,...,2,Personal Auto,Personal L3,Offer3,Call Center,480.000000,SUV,Medsize,A,2
3,XL78013,Oregon,22332.439460,Yes,Extended,College,2011-01-11,Employed,M,49078,...,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A,1
4,QA50777,Oregon,9025.067525,No,Premium,Bachelor,2011-01-17,Medical Leave,F,23675,...,7,Personal Auto,Personal L2,Offer1,Branch,707.925645,Four-Door Car,Medsize,A,1


In [63]:
df["sales_channel"].value_counts()

sales_channel
Agent          4121
Branch         3022
Call Center    2141
Web            1626
Name: count, dtype: int64

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [67]:
df.columns = df.columns.str.lower().str.replace(" ", "_")

In [71]:
channels = ["branch", "call_center", "web", "mail"]

df_filtered = df[df["sales_channel"].isin(channels)]

In [72]:
revenue_pivot = df.pivot_table(
    index="sales_channel",
    values="total_claim_amount",
    aggfunc="sum"
).round(2).sort_values(by="total_claim_amount", ascending=False)

revenue_pivot

,total_claim_amount
sales_channel,
Agent,1810226.82
Branch,1301204.00
Call Center,926600.82
Web,706600.04


El canal de Agentes (Agent) es el más efectivo, generando 1,810,226.82 en ingresos totales. Este resultado indica que la interacción directa con agentes es el principal motor de ventas y generación de valor para la empresa.

El canal de Sucursal (Branch) ocupa el segundo lugar con 1,301,204.00, mostrando un rendimiento sólido, aunque significativamente inferior al canal de agentes.

El Call Center genera 926,600.82, lo que representa una contribución intermedia. Aunque sigue siendo relevante, su eficiencia es menor en comparación con los canales presenciales.

Por último, el canal Web es el que menor ingresos genera con 706,600.04, lo que sugiere un bajo rendimiento del canal digital en comparación con los canales tradicionales.

- nota : El enunciado menciona el canal “mail”, pero este no aparece en el dataset. En su lugar, el dataset incluye el canal “Agent”, por lo que el análisis se ha realizado utilizando las categorías reales disponibles en los datos.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [65]:
df["customer_lifetime_value"].dtype

dtype('float64')

In [66]:
df["customer_lifetime_value"] = pd.to_numeric(
    df["customer_lifetime_value"],
    errors="coerce"
)

In [73]:
clv_pivot = df.pivot_table(
    index=["gender", "education"],
    values="customer_lifetime_value",
    aggfunc="mean"
).round(2)

clv_pivot.unstack()

customer_lifetime_value                                         \
education                Bachelor  College   Doctor High School or Below   
gender                                                                     
F                         7874.27  7748.82  7328.51              8675.22   
M                         7703.60  8052.46  7415.33              8149.69   

                    
education   Master  
gender              
F          8157.05  
M          8168.83

El Customer Lifetime Value (CLV) varía ligeramente entre segmentos de género y nivel educativo, aunque no se observan diferencias extremadamente grandes.

En el caso de las mujeres (F), el mayor CLV se encuentra en el nivel educativo de High School or Below (8675.22), seguido de Master (8157.05). En contraste, el valor más bajo aparece en el nivel de Doctor (7328.51).

Para los hombres (M), los valores de CLV son relativamente estables, destacando ligeramente Master (8168.83) y College (8052.46) como los segmentos más altos. El valor más bajo se observa en Bachelor (7703.60).

Las diferencias en CLV entre los segmentos son relativamente pequeñas, lo que sugiere que no hay una segmentación fuerte por género o nivel educativo. Por tanto, genero y nivel educativo parece ser un factor determinante fuerte. La empresa podría obtener mejores resultados enfocándose en otros factores como el canal de venta o el tipo de póliza.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [ ]:
# Your code goes here